# Build a regression model.

In [42]:
import pandas as pd
import numpy as np
import requests
import datetime
import statsmodels.api as sm
import seaborn as sns
import matplotlib.pyplot as plt
import os
%matplotlib inline 
sns.set(color_codes=True)

# Foursquare
Provide model output and an interpretation of the results. 

In [44]:
df_bikes_fsq = pd.read_csv("../data/bikes_fsq_data.csv", index_col=0)
display(df_bikes_fsq.head(5))

empty_slots  free_bikes                        Station_ID  \
0           19          16  7a19c49f486d7c0c02b3685d7b240448   
1           19          16  7a19c49f486d7c0c02b3685d7b240448   
2           19          16  7a19c49f486d7c0c02b3685d7b240448   
3           19          16  7a19c49f486d7c0c02b3685d7b240448   
4           19          16  7a19c49f486d7c0c02b3685d7b240448   

   station_latitude  station_longitude        station_name  extra.slots  \
0         49.262487        -123.114397  0001 10th & Cambie           35   
1         49.262487        -123.114397  0001 10th & Cambie           35   
2         49.262487        -123.114397  0001 10th & Cambie           35   
3         49.262487        -123.114397  0001 10th & Cambie           35   
4         49.262487        -123.114397  0001 10th & Cambie           35   

  extra.status        Station Name                    FSQ_ID  ...   Longitude  \
0       online  0001 10th & Cambie  51b201d27dd249ae714ba728  ... -123.105307   
1       online  0001 10th & Cambie  593c332b16fa0435e42fb0d5  ... -123.114269   
2       online  0001 10th & Cambie  4aab378af964a520445920e3  ... -123.105073   
3       online  0001 10th & Cambie  5177156c498ea4fd6b901dda  ... -123.110286   
4       online  0001 10th & Cambie  4aa8320cf964a520f94f20e3  ... -123.115731   

   Distance                                            Address  \
0       674  15 8th Ave W (btwn Manitoba & Ontario), Vancou...   
1        93            467 West Broadway, Vancouver BC V5Y 1R4   
2       700  3 8th Ave W (at Ontario St), Vancouver BC V5Y 1M8   
3       299  243 West Broadway (at Alberta St), Vancouver B...   
4       144  521W Broadway W (at Cambie St), Vancouver BC V...   

         Category name Rating Popularity  Price  Total Photos  Total Ratings  \
0                  Bar    9.2   0.959464    2.0         590.0          388.0   
1  American Restaurant    6.4   0.978858    1.0          17.0            7.0   
2         Organization    6.9   0.877072    1.0          33.0           40.0   
3                 Café    8.4   0.923845    1.0          68.0           96.0   
4                 Café    5.4   0.977400    1.0          33.0           31.0   

   Total Tips  
0        75.0  
1         3.0  
2        10.0  
3        30.0  
4        15.0  

[5 rows x 22 columns]

In [45]:
df_bikes_fsq.dtypes

empty_slots            int64
free_bikes             int64
Station_ID            object
station_latitude     float64
station_longitude    float64
station_name          object
extra.slots            int64
extra.status          object
Station Name          object
FSQ_ID                object
Venue Name            object
Latitude             float64
Longitude            float64
Distance               int64
Address               object
Category name         object
Rating               float64
Popularity           float64
Price                float64
Total Photos         float64
Total Ratings        float64
Total Tips           float64
dtype: object

In [46]:
y = df_bikes_fsq['Total Photos']
X = df_bikes_fsq['free_bikes']
X = sm.add_constant(X) #adds a column of 1's so the model will contain an intercept

model = sm.OLS(y, X) #instantiate
results = model.fit() #fit the model (this is where OLS is actually being run)
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:           Total Photos   R-squared:                       0.005
Model:                            OLS   Adj. R-squared:                  0.004
Method:                 Least Squares   F-statistic:                     19.58
Date:                Tue, 08 Nov 2022   Prob (F-statistic):           9.87e-06
Time:                        02:00:20   Log-Likelihood:                -27471.
No. Observations:                4285   AIC:                         5.495e+04
Df Residuals:                    4283   BIC:                         5.496e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         85.6531      4.080     20.995      0.0

I created a Simple Linear Regression. As seen above, the adjusted R-squared is 0.004. With a p-value as 0, the variable is significant in the linear regression. So there is almost no correlation between the number of total bikes and the number of photos taken at the POI location but the variable is significant in the model. 



In [47]:
y = df_bikes_fsq['Popularity']
X = df_bikes_fsq['free_bikes']
X = sm.add_constant(X) #adds a column of 1's so the model will contain an intercept

model = sm.OLS(y, X) #instantiate
results = model.fit() #fit the model (this is where OLS is actually being run)
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:             Popularity   R-squared:                       0.003
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     12.24
Date:                Tue, 08 Nov 2022   Prob (F-statistic):           0.000472
Time:                        02:00:22   Log-Likelihood:                 3527.9
No. Observations:                4285   AIC:                            -7052.
Df Residuals:                    4283   BIC:                            -7039.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.9131      0.003    310.236      0.0

As seen above, the adjusted R-squared is 0.003. With a p-value as 0, the variable is significant in the linear regression. So there is almost no correlation between the number of total bikes and the popularity at the POI location, but the variable is significant in the model. 

In [48]:
y = df_bikes_fsq['Total Ratings']
X = df_bikes_fsq['extra.slots'] # total number of bike slots
X = sm.add_constant(X) #adds a column of 1's so the model will contain an intercept

model = sm.OLS(y, X) #instantiate
results = model.fit() #fit the model (this is where OLS is actually being run)
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:          Total Ratings   R-squared:                       0.002
Model:                            OLS   Adj. R-squared:                  0.002
Method:                 Least Squares   F-statistic:                     8.741
Date:                Tue, 08 Nov 2022   Prob (F-statistic):            0.00313
Time:                        02:00:24   Log-Likelihood:                -25700.
No. Observations:                4285   AIC:                         5.140e+04
Df Residuals:                    4283   BIC:                         5.142e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const          63.4876      5.546     11.447      

As seen above, the adjusted R-squared is 0.002. With a p-value as 0.003, the variable is significant in the linear regression. There is almost no correlation between the total number of total bike slots and the total number of ratings at the POI location.

# Yelp

Provide model output and an interpretation of the results. 

In [49]:
df_yelp = pd.read_csv("../data/yelp_restaurant_bar_cafe_data.csv", index_col=0)
display(df_yelp.head(5))

Station ID        Station Name  \
0  7a19c49f486d7c0c02b3685d7b240448  0001 10th & Cambie   
1  7a19c49f486d7c0c02b3685d7b240448  0001 10th & Cambie   
2  7a19c49f486d7c0c02b3685d7b240448  0001 10th & Cambie   
3  7a19c49f486d7c0c02b3685d7b240448  0001 10th & Cambie   
4  7a19c49f486d7c0c02b3685d7b240448  0001 10th & Cambie   

                  Yelp ID                            Venue Name Price  \
0  PrKpa6Ejv1v1WvIIywYMvg                   Aperture Coffee Bar    $$   
1  NensKn1MSVU_rm-1Y6WlFA                         Marulilu Cafe    $$   
2  X-73k3cwBjjc170MOdPa7A                        Elysian Coffee    $$   
3  dh-h2jbsa2ElN5Db0vfotQ  On the Grind - Pedalling Epic Coffee     $   
4  UkSNdIF54dvas25eh_CmFw                      Caffe Cittadella    $$   

   Review Count  Rating   Latitude   Longitude    Distance  \
0           122     4.0  49.263204 -123.110470  295.924920   
1           281     4.0  49.263380 -123.114150  102.329817   
2           207     3.5  49.262940 -123.117280  221.080646   
3             4     5.0  49.261226 -123.113927  144.308422   
4           152     4.0  49.264809 -123.117267  331.734686   

                                          Address Category name  
0   243 W Broadway, Vancouver, BC V5Y 3L1, Canada  Coffee & Tea  
1   451 W Broadway, Vancouver, BC V5Y 1R4, Canada         Cafes  
2   590 W Broadway, Vancouver, BC V5Z 1E9, Canada  Coffee & Tea  
3                           Vancouver, BC, Canada  Coffee & Tea  
4  2310 Ash Street, Vancouver, BC V5Z 3C2, Canada  Coffee & Tea

In [52]:
y = df_bikes_fsq['Price']
X = df_bikes_fsq['free_bikes']
X = sm.add_constant(X) #adds a column of 1's so the model will contain an intercept

model = sm.OLS(y, X) #instantiate
results = model.fit() #fit the model (this is where OLS is actually being run)
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  Price   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.000
Method:                 Least Squares   F-statistic:                    0.4064
Date:                Tue, 08 Nov 2022   Prob (F-statistic):              0.524
Time:                        08:54:27   Log-Likelihood:                -4112.7
No. Observations:                4285   AIC:                             8229.
Df Residuals:                    4283   BIC:                             8242.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5337      0.018     87.605      0.0

As seen above, the adjusted R-squared is 0.002. With a p-value as 0.524, the variable is not significant in the linear regression. There is almost no correlation between the total number of total free bikes and the price at the POI location.

In [53]:
y = df_bikes_fsq['Rating']
X = df_bikes_fsq['free_bikes']
X = sm.add_constant(X) #adds a column of 1's so the model will contain an intercept

model = sm.OLS(y, X) #instantiate
results = model.fit() #fit the model (this is where OLS is actually being run)
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                 Rating   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                  0.000
Method:                 Least Squares   F-statistic:                     1.789
Date:                Tue, 08 Nov 2022   Prob (F-statistic):              0.181
Time:                        08:55:45   Log-Likelihood:                -6006.7
No. Observations:                4285   AIC:                         1.202e+04
Df Residuals:                    4283   BIC:                         1.203e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.1780      0.027    263.526      0.0

As seen above, the adjusted R-squared is 0. There is no correlation between the total number of total free bikes and the price at the POI location. With a p-value as 0.181, the variable is not considered significant in the linear regression. 

# Stretch

How can you turn the regression model into a classification model?

I didn't have enough time, but try this when I can.